### ISSUE TITLE
Error in installation of Trilinos

### ISSUE THREAD

**Issue by AjaySingh40** [trilinos/Trilinos#12873]
/home../trilinos/Trilinos-trilinos-release-15-1-1/packages/seacas/libraries/ioss/src/main/io_modify.C: In function ‘bool {anonymous}::handle_rename(const std::vector<std::__cxx11::basic_string<char> >&, Ioss::Region&)’:
/home../trilinos/Trilinos-trilinos-release-15-1-1/packages/seacas/libraries/ioss/src/main/io_modify.C:1134:7: error: ‘ex_entity_type’ was not declared in this scope; did you mean ‘get_entity_type’?
 1134 |       ex_entity_type exo_type  = Ioex::map_exodus_type(ioss_type);
      |       ^~~~~~~~~~~~~~
      |       get_entity_type
/home../trilinos/Trilinos-trilinos-release-15-1-1/packages/seacas/libraries/ioss/src/main/io_modify.C:1135:53: error: ‘exo_type’ was not declared in this scope
 1135 |       auto           ierr      = ex_put_name(exoid, exo_type, id(ge), new_name.c_str());
      |                                                     ^~~~~~~~
/home../trilinos/Trilinos-trilinos-release-15-1-1/packages/seacas/libraries/ioss/src/main/io_modify.C:1135:34: error: ‘ex_put_name’ was not declared in this scope
 1135 |       auto           ierr      = ex_put_name(exoid, exo_type, id(ge), new_name.c_str());
      |                                  ^~~~~~~~~~~
/home../trilinos/Trilinos-trilinos-release-15-1-1/packages/seacas/libraries/ioss/src/main/io_modify.C:1136:19: error: ‘EX_NOERR’ was not declared in this scope
 1136 |       if (ierr != EX_NOERR) {
      |                   ^~~~~~~~
/home../trilinos/Trilinos-trilinos-release-15-1-1/packages/seacas/libraries/ioss/src/main/io_modify.C:1137:9: error: ‘Ioex’ has not been declared
 1137 |         Ioex::exodus_error(exoid, __LINE__, __func__, __FILE__);

/home../trilonos/Trilinos-trilinos-release-15-1-1/packages/seacas/libraries/ioss/src/main/io_modify.C: In function ‘void {anonymous}::modify_time(Ioss::Region&, double, double)’:
/home../trilinos/Trilinos-trilinos-release-15-1-1/packages/seacas/libraries/ioss/src/main/io_modify.C:1871:5: error: ‘ex_get_all_times’ was not declared in this scope
 1871 |     ex_get_all_times(exoid, times.data());
      |     ^~~~~~~~~~~~~~~~
/home../trilinos/Trilinos-trilinos-release-15-1-1/packages/seacas/libraries/ioss/src/main/io_modify.C:1875:7: error: ‘ex_put_time’ was not declared in this scope
 1875 |       ex_put_time(exoid, step + 1, &times[step]);
      |       ^~~~~~~~~~~
[100%] Linking CXX executable shell_to_hex
[100%] Built target shell_to_hex
[100%] Linking CXX executable cth_pressure_map
make[2]: *** [packages/seacas/libraries/ioss/src/main/CMakeFiles/io_modify.dir/io_modify.C.o] Error 1
make[2]: *** Waiting for unfinished jobs....
[100%] Built target cth_pressure_map
make[1]: *** [packages/seacas/libraries/ioss/src/main/CMakeFiles/io_modify.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[100%] Linking CXX executable io_info
[100%] Built target io_info
[100%] Linking CXX executable sphgen
[100%] Linking CXX executable exomatlab
[100%] Linking CXX executable io_shell
[100%] Built target sphgen
[100%] Built target exomatlab
[100%] Linking CXX executable skinner
[100%] Built target io_shell
[100%] Built target skinner
make: *** [all] Error 2

**Comment by cgcgcg** [comment]
@gsjaardema Could this be related to #12870?

**Comment by ikalash** [comment]
We are seeing Exodus related issues in the Albany nightlies too.  Please see, e.g., https://sems-cdash-son.sandia.gov/cdash/viewBuildError.php?buildid=62079 .

**Comment by gsjaardema** [comment]
There was another version of this on github.io that I responded to:

I think there is an incorrect ifdef guarding whether exodus is/isn't enabled. Since other exodus-related codes are being built (exomatlab, sphgen, skinner...) it looks like exodus is enabled, but SEACAS_HAVE_EXODUS is not being defined. I will check to see where my logic fails.

I guess I should ask explicitly -- does this build have Exodus enabled?

**Comment by gsjaardema** [comment]
> @gsjaardema Could this be related to #12870?

Probably, but I'm not sure why/how yet...

**Comment by gsjaardema** [comment]
> We are seeing Exodus related issues in the Albany nightlies too. Please see, e.g., https://sems-cdash-son.sandia.gov/cdash/viewBuildError.php?buildid=62079 .

Is this (intel-19.0) the only build having issues, or are several?  Not sure if it is a C++17 compatability issue or where to look yet...

**Comment by ikalash** [comment]
@gsjaardema : looking more carefully at our nightlies, the URL I pointed you is for an intel 19 build on attaway.  However, this is not just an intel thing b/c I am seeing it on my RHEL8 workstation, camobap, which use gcc (they don't post to our CDash site for Trilinos though).  Looking at that output, though, it appears the error is different:
```
packages/seacas/libraries/exodus/libexodus.so.15.2.0: undefined reference to `nc_def_var_fill'
collect2: error: ld returned 1 exit status
```
I can provide the full log if you are interested.  Camobap uses gcc-11.1.1.  

A full list of our nightlies are here: https://sems-cdash-son.sandia.gov/cdash/index.php?project=Albany .  Most of them are not seeing issues for Trilinos.

**Comment by ikalash** [comment]
@gsjaardema : also, it seems there is an obsolete header included in one of the Ioss files.  It shows up in the Albany builds for which Trilinos succeeds: https://sems-cdash-son.sandia.gov/cdash/viewBuildError.php?buildid=62647 .  

```
/home/ikalash/Trilinos_Albany/nightlyAlbanyTests/Results/Trilinos/build-no-epetra/install/include/Ioss_Region.h:26:10: fatal error: fmt/ostream.h: No such file or directory
```

**Comment by gsjaardema** [comment]
> @gsjaardema : also, it seems there is an obsolete header included in one of the Ioss files. It shows up in the Albany builds for which Trilinos succeeds: https://sems-cdash-son.sandia.gov/cdash/viewBuildError.php?buildid=62647 .
> 
> ```
> /home/ikalash/Trilinos_Albany/nightlyAlbanyTests/Results/Trilinos/build-no-epetra/install/include/Ioss_Region.h:26:10: fatal error: fmt/ostream.h: No such file or directory
> ```

That file should be at `../packages/seacas//libraries/ioss/src/private_copy_fmt/fmt/ostream.h`

**Comment by ikalash** [comment]
@gsjaardema : it is there, I can confirm, but it does not seem to be copied over to the install/include directory.  Did something change with the build rules, somehow?

**Comment by gsjaardema** [comment]
Yes, I just saw that... It is not failing as part of a Trilinos build, it is failing when Albany is trying to use a Trilinos install.  The embeded fmt used in Trilinos only works during the Trilinos build and won't be found when Ioss_Region.h is used in another package...   (In the actual SEACAS installs, we are using an external fmt library, so everything works..)

I will protect the include of `fmt/ostream.h` so not needed in Ioss_Region.h for Trilinos.  Sorry for the problem.

**Comment by gsjaardema** [comment]
See #12875 for fix of `fmt/ostream.h` issue.  I think that is the only place that causes this problem.  Will need longer-term fix.

**Comment by ikalash** [comment]
> What version of the netCDF library are you using and do you have the netcdf-4 capability enabled?

On camobap (where there is the undefined error is encountered), it looks like I am using netcdf-4.4.1.  It is build from source code on that machine using this configure script: https://github.com/sandialabs/Albany/blob/master/doc/nightlyTestHarness/camobap/netcdf-4.4.1-configure.sh .  It looks like netcdf-4 is disabled.

**Comment by ikalash** [comment]
> See #12875 for fix of `fmt/ostream.h` issue. I think that is the only place that causes this problem. Will need longer-term fix.

Thank you!  I will try this shortly.

**Comment by gsjaardema** [comment]
Sea also #12876 for PR addressing the netCDF `nc_def_var_fill` issue.

**Comment by gsjaardema** [comment]
See #12877 for fixes that are correctly on develop and not master.

**Comment by cgcgcg** [comment]
@gsjaardema Since it appears that #12877 fixes the issues: how can we prevent this from happening in the future? Is there an issue regarding testing against TPLs? Or is this a gap in test coverage of SEACAS?

**Comment by gsjaardema** [comment]
I need to add testing the build with earlier versions of the netCDF library and with multiple configurations (with/without HDF5, pNetCDF, parallel).  We have some of that in place, but it obviously missed this change.  I will make sure I have a test build that uses an earlier version of the netCDF library.

**Comment by gsjaardema** [comment]
One other issue is that Trilinos does not currently provide `fmt` as a TPL, so we rely on an internal embedded version.  Doing the snapshot into Trilinos correctly such that we don't lose the embedded fmt that exists in Trilinos/SEACAS, but not external/SEACAS needs some care.  The embedded version also means that the users of SEACAS include files will probably not have access to the fmt/xyz.h include files, since the embedded fmt include files are currently not installed...

**Comment by ikalash** [comment]
Everything is good now in the Albany nightlies.  Sorry for not commenting earlier - there were a lot of issues open and I lost track of them a bit...

**Comment by gsjaardema** [comment]
> Everything is good now in the Albany nightlies. Sorry for not commenting earlier - there were a lot of issues open and I lost track of them a bit...

Not a problem, I operate under the "no news is good news" maxim

**Comment by cgcgcg** [comment]
Would it make sense to add fmt as a TPL? Would that overall reduce the amount of testing that needs to happen? It's a pretty small library without any dependencies, right?

**Comment by ccober6** [comment]
> Would it make sense to add fmt as a TPL? Would that overall reduce the amount of testing that needs to happen? It's a pretty small library without any dependencies, right?
> 
> EDIT: Looks like fmt is available from spack. So it probably is easy on the framework team to add that to our environments.

Yeah, I had the same question.  @browne, I assume this would not be too hard.  If so, let's get leadership input and possibly included.

**Comment by ccober6** [comment]
miss tagged @sebrowne .  :(

**Comment by rppawlo** [comment]
fmt is included in the c++20 standard. Unless you need something that is not in the std version, waiting for the minimum standard requirement to change to c++20 next FY is also an option.

**Comment by cgcgcg** [comment]
I think that's a good option.

**Comment by sebrowne** [comment]
External `fmt` is the predominant use case for SEACAS, I'm perfectly happy to change the containerized (AT2) builds to use it from Spack in the interim (this should give us test coverage fairly soon), and then once we move to C++20 it becomes moot.

**Comment by gsjaardema** [comment]
> fmt is included in the c++20 standard. Unless you need something that is not in the std version, waiting for the minimum standard requirement to change to c++20 next FY is also an option.

True, but I also need Sierra to move to C++20.  Will probably need to figure out how to have them coexist which might be possible...
